Source/Dafny/Verifier/Translator.ExpressionTranslator.cs

```diff
@@ -1422,17 +1422,27 @@ public ExpressionTranslator(ExpressionTranslator etran, string modifiesFrame)
         }
       }
 
+      // If the LHS has type Bitvector(w), cast the RHS to the number of bits
+      // required to represent w - 1 and then to Bitvector(w), to result in
+      // an easier SMT problem. This is always valid because the value of the
+      // RHS must be less than w.
+      private Expr TrShiftRotateExpr(string fnName, Expression expr, Expr lhs, Expr rhs, Type rhsType, int w, bool liftLit) {
+        var intermediateBvWidth = (int)(new BigInteger(w - 1)).GetBitLength();
+        translator.AddBitwidth(intermediateBvWidth);
+        var tok = GetToken(expr);
+        var rhsConvertedType = new BitvectorType(intermediateBvWidth);
+        var rhsConverted = translator.ConvertExpression(tok, rhs, rhsType, rhsConvertedType);
+        var rhsFinal = translator.ConvertExpression(tok, rhsConverted, rhsConvertedType, expr.Type);
+        return TrToFunctionCall(tok, fnName + "_bv" + w, translator.BplBvType(w), lhs, rhsFinal, liftLit);
```

Comment:
Why does the RHS need to be the same type as the LHS in Boogie?

Also, what happens if the RHS value doesn't fit into the intermediate type? Is there a check someone else checks constrains this value? Can we refer to that check here to indicate why this is safe.

Reply:
The RHS needs to be the same type as the LHS because Boogie's bit vector operations are exactly those in SMT-Lib, and the [SMT-Lib standard](https://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml) constrains them that way. I'm not sure what the rationale is behind it.

I'm glad you brought up whether the value fits in the intermediate type, though! I had thought that this was being checked elsewhere, and it sort of is, but there's actually an off-by-one between that check and what I need here. 

Dafny already emits a check that you don't shift (or rotate) more than the bit vector width. But you **can** shift by exactly the bit vector width (turning it into all zeros), and that one value won't fit into the bit vector of the size used here. So it needs to be one more bit.